openpype/hosts/maya/plugins/publish/extract_xgen.py

```diff
@@ -0,0 +1,158 @@
+import os
+import copy
+import tempfile
+
+from maya import cmds
+import xgenm
+
+from openpype.pipeline import publish
+from openpype.hosts.maya.api.lib import (
+    maintained_selection, attribute_values, write_xgen_file
+)
+from openpype.lib import StringTemplate
+
+
+class ExtractXgenCache(publish.Extractor):
+    """Extract Xgen
+
+    Workflow:
+    - Duplicate nodes used for patches.
+    - Export palette and import onto duplicate nodes.
+    - Export/Publish duplicate nodes and palette.
+    - Export duplicate palette to .xgen file and add to publish.
+    - Publish all xgen files as resources.
+    """
+
+    label = "Extract Xgen"
+    hosts = ["maya"]
+    families = ["xgen"]
+    scene_type = "ma"
+
+    def process(self, instance):
+        if "representations" not in instance.data:
+            instance.data["representations"] = []
+
+        staging_dir = self.staging_dir(instance)
+        maya_filename = "{}.{}".format(instance.data["name"], self.scene_type)
+        maya_filepath = os.path.join(staging_dir, maya_filename)
+
+        # Collect nodes to export.
```

Comment:
How would you clean up a failed scene change, if you dont know what nodes/files to clean up?

Reply:
In some case it worked magically if I did:

```python
with undo_chunk():
    # Do all your changes here

maya.cmds.undo()
```
Hehe. But I'd say that if you do complex things that might be unsafe - however, if all commands you are doing are undoable I wouldn't know why this would fail.

Other ways would be to e.g. use the [delete_after](https://github.com/ynput/OpenPype/blob/2d574996c38e7990eb66c1dd4a98d8be30f5703c/openpype/hosts/maya/api/lib.py#L639-L676) context manager where you add the nodes you are generating as soon as possible after you've generated them so that IF an errors gets raised at least those nodes would get cleaned up.

It seems that in this case the `delete_after` workflow might be possible for your code?

Reply:
Thanks for pointing me to `delete_after`. Will use that.